<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [102]:
import pandas as pd
import psycopg2

In [104]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [105]:
# текст запроса
query_3_1 = f'''SELECT count(*) count_vacansies
                FROM vacancies
            '''

In [106]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
display(df)
print('Количество вакансий:', df_3_1['count_vacansies'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\3658164851.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection)


count_vacansies
0            49197

Количество вакансий: 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [107]:
# текст запроса
query_3_2 = f'''SELECT count(*) count_employers
                FROM employers
            '''

In [108]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
display(df_3_2)
print('Количество работодателей:', df_3_2['count_employers'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2455156735.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


count_employers
0            23501

Количество работодателей: 23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [109]:
# текст запроса
query_3_3 = f'''SELECT count(*) count_areas
                FROM areas
            '''

In [110]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
display(df_3_3)
print('Количество регионов:', df_3_3['count_areas'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\1544855318.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


count_areas
0         1362

Количество регионов: 1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [111]:
# текст запроса
query_3_4 = f'''SELECT count(*) count_industries
                FROM industries
            '''

In [112]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
display(df_3_4)
print('Количество сфер деятельности:', df_3_4['count_industries'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2074271497.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


count_industries
0               294

Количество сфер деятельности: 294


***

В базе имеется 49197 вакансий и 23501 работодателей из 1362 регионов в 294 сферах деятельности

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [113]:
# текст запроса
query_4_1 = f'''SELECT 
                    a.name area_name,
                    count(v) count_vacancies
                FROM areas a
                JOIN vacancies v ON a.id = v.area_id
                GROUP BY a.id
                ORDER BY 2 DESC
            '''


In [114]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
display(df_4_1)

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\3084987771.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


area_name  count_vacancies
0                      Москва             5333
1             Санкт-Петербург             2851
2                       Минск             2112
3                 Новосибирск             2006
4                      Алматы             1892
..                        ...              ...
764                    Кизляр                1
765                    Джизак                1
766                    Эртиль                1
767                  Арсеньев                1
768  Тура (Красноярский край)                1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [115]:
# текст запроса
query_4_2 = f'''SELECT count(*)
                FROM vacancies
                WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
            '''

In [116]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
display(df_4_2)
print('Количество вакансий, у которых заполнено хотя бы одно из двух полей с зарплатой:', df_4_2['count'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2312388760.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


count
0  24073

Количество вакансий, у которых заполнено хотя бы одно из двух полей с зарплатой: 24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [117]:
# текст запроса
query_4_3 = f'''SELECT 
                    salary_from,
                    salary_to
                FROM vacancies
            '''

In [118]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
print('Среднее значение нижней границы зарплатной вилки:', round(df_4_3['salary_from'].mean()), '\n',
      'Cреднее значение верхней границы зарплатной вилки:', round(df_4_3['salary_to'].mean()))

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\842239505.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


Среднее значение нижней границы зарплатной вилки: 71065 
 Cреднее значение верхней границы зарплатной вилки: 110537


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [119]:
# текст запроса
query_4_4 = f'''SELECT 
                    schedule,
                    employment,
                    count(*) count_vacancies
                FROM vacancies 
                GROUP BY schedule, employment
                ORDER BY count(*) DESC
            '''

In [120]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
display(df_4_4)

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\972204354.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


schedule           employment  count_vacancies
0        Полный день     Полная занятость            35367
1   Удаленная работа     Полная занятость             7802
2      Гибкий график     Полная занятость             1593
3   Удаленная работа  Частичная занятость             1312
4     Сменный график     Полная занятость              940
5        Полный день           Стажировка              569
6     Вахтовый метод     Полная занятость              367
7        Полный день  Частичная занятость              347
8      Гибкий график  Частичная занятость              312
9        Полный день     Проектная работа              141
10  Удаленная работа     Проектная работа              133
11     Гибкий график           Стажировка              116
12    Сменный график  Частичная занятость              101
13  Удаленная работа           Стажировка               64
14     Гибкий график     Проектная работа               18
15    Сменный график           Стажировка               12
16    Вахтовый метод     Проектная работа                2
17    Сменный график     Проектная работа                1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [121]:
# текст запроса
query_4_5 = f'''SELECT 
                    experience,
                    count(*) count_vacancies
                FROM vacancies 
                GROUP BY experience
                ORDER BY count(*)
            '''

In [122]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
display(df_4_5)

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\1393685526.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


experience  count_vacancies
0         Более 6 лет             1337
1           Нет опыта             7197
2       От 3 до 6 лет            14511
3  От 1 года до 3 лет            26152

***

Из анализа работодателей можно сделать выводы:
1. Больше всего вакансий в Москве - 5333, связано это с тем, что Москва - столица Российской Федерации, а также самый крупный город в стране;
2. Из 49197 вакансий заполнено хотя бы одно из полей с заработной платой, лишь у 24073;
3. Среднее значение нижней границы зарплатной вилки составяет 71065, а верхней границы зарплатной вилки - 110537;
4. Наиболее востребованный график работы и тип занятости - полный день и полная занятость соотвественно, 35367 вакансий на платформе;
5. Наибольшее количество вакансий с оптытом работы от 1 года до 3 лет, их колчисетво - 26152

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [123]:
# текст запроса
query_5_1 = f'''-- Первый запрос для определния работодателя-лидера по колчисетву вакансий --
                (SELECT 
                    e.name employer,
                    count(v) count_vacancies
                FROM employers e
                JOIN vacancies v ON e.id = v.employer_id
                GROUP BY e.name
                ORDER BY count(v) DESC
                LIMIT 1)
                UNION
                -- Второй запрос для определения работодателя на пятом месте по количеству вакансий --
                (SELECT 
                    ee.name,
                    count(vv)
                FROM employers ee
                JOIN vacancies vv ON ee.id = vv.employer_id
                GROUP BY ee.name
                ORDER BY count(vv) DESC
                OFFSET 4
                LIMIT 1)
                ORDER BY 2 DESC
            '''

In [124]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
display(df_5_1)
print('Работодатели находятся на первом и пятом месте по количеству вакансий: 1 -', df_5_1['employer'][0],', 5 -', df_5_1['employer'][1])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\1829321321.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


employer  count_vacancies
0         Яндекс             1933
1  Газпром нефть              331

Работодатели находятся на первом и пятом месте по количеству вакансий: 1 - Яндекс , 5 - Газпром нефть


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [125]:
# текст запроса
query_5_2 = f'''SELECT
                    a.name area,
                    count(e) count_employers,
                    count(v) count_vacancies
                FROM areas a
                LEFT JOIN employers e ON a.id = e.area
                LEFT JOIN vacancies v ON a.id = v.area_id
                GROUP BY a.name
                HAVING count(v) = 0
                ORDER BY 2 DESC
                LIMIT 1
            '''

In [126]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
print('Регион, в котором нет вакансий и с наибольшим количеством работодателей:', df_5_2['area'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\1949925699.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


Регион, в котором нет вакансий и с наибольшим количеством работодателей: Россия


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [127]:
# текст запроса
query_5_3 = f'''SELECT
                    distinct e.name employer,
                    count(distinct v.area_id) count_areas
                FROM employers e
                LEFT JOIN vacancies v ON e.id = v.employer_id
                GROUP BY e.id
                ORDER BY 2 DESC
            '''

In [128]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
display(df_5_3)

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\1820345300.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


employer  count_areas
0                       Яндекс          181
1                   Ростелеком          152
2                   Спецремонт          116
3       Поляков Денис Иванович           88
4                    ООО ЕФИН            71
...                        ...          ...
23336             Ярпож Казань            0
23337                   ЯРСНИП            0
23338   Ясли-сад № 28 г.Минска            0
23339            Яшин&Партнёры            0
23340  Яшнова Алена Викторовна            0

[23341 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [129]:
# текст запроса
query_5_4 = f'''SELECT count(*)
                FROM employers e
                LEFT JOIN employers_industries ei on e.id = ei.employer_id
                WHERE ei.industry_id is null
            '''

In [130]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
display(df_5_4)
print('Количество работодателей не указавших сферу деятельности:', df_5_4['count'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\652995106.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


count
0   8419

Количество работодателей не указавших сферу деятельности: 8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [131]:
# текст запроса
query_5_5 = f'''SELECT
                    distinct e.name
                FROM employers e
                LEFT JOIN employers_industries ei on e.id = ei.employer_id
                GROUP BY e.id
                HAVING count(ei.industry_id) = 4
                ORDER BY e.name
                OFFSET 2
                LIMIT 1
            '''

In [132]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
display(df_5_5)
print('Название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности:', df_5_5['name'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2975771326.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

Название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности: 2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [133]:
# текст запроса
query_5_6 = f'''SELECT
                    count(ei.employer_id)
                FROM employers_industries ei
                LEFT JOIN industries i ON i.id = ei.industry_id
                GROUP BY i.id
                HAVING i.name = 'Разработка программного обеспечения'
            '''

In [134]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
display(df_5_6)
print('Количество работодателей, у которых в качестве сферы деятельности указана Разработка программного обеспечения:', df_5_6['count'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2201635212.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

Количество работодателей, у которых в качестве сферы деятельности указана Разработка программного обеспечения: 3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [135]:
# код для получения списка городов-милионников
from bs4 import BeautifulSoup
import requests

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' 
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
data = []
HTML_data = page.find_all('table')[0].find_all('tr')[1:]
for element in HTML_data:
    sub_data = []
    for i, k in enumerate(element):
        try:
            if i == 1:
                sub_data.append(k.get_text())
        except:
            continue
    data.append(sub_data)
data_result = []
for elem in data:
    data_result.extend(elem)
result = tuple(data_result)
print(result)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [136]:
# текст запроса
query_5_7 = f''' -- Первый запрос для определния количества вакансий в городах-миллионниках --
               SELECT
                    a.name city,
                    count(v.id) count_vacancies
               FROM areas a 
               LEFT JOIN vacancies v ON v.area_id = a.id
               LEFT JOIN employers e ON e.id = v.employer_id
               WHERE e.name = 'Яндекс' AND a.name IN {result}
               GROUP BY a.name 
               UNION ALL
               -- Второй запрос для определения общего количества вакансий во всех города-миллионниках -- 
               SELECT 
                    'Total',
                    count(v.id)
               FROM areas a 
               LEFT JOIN vacancies v ON v.area_id = a.id
               LEFT JOIN employers AS e ON e.id = v.employer_id
               WHERE e.name = 'Яндекс' AND a.name IN {result}
               GROUP BY 1
               ORDER BY 2
              '''

In [137]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
display(df_5_7)

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\796585636.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


city  count_vacancies
0              Омск               21
1         Челябинск               22
2        Красноярск               23
3         Волгоград               24
4             Пермь               25
5            Казань               25
6    Ростов-на-Дону               25
7               Уфа               26
8            Самара               26
9         Краснодар               30
10          Воронеж               32
11      Новосибирск               35
12  Нижний Новгород               36
13     Екатеринбург               39
14  Санкт-Петербург               42
15           Москва               54
16            Total              485

***

Из анализа работодателей можно сделать выводы:
1. Больше всего вакансий имеется у Яндекс - 1933 вакансии на платформе;
2. Работодатель, который публикует вакансии в 181 регионе - Яндекс. Он является лидером по покрытию регионов вакансиями;
3. 8419 работодателей не указали сферу деятельности;
4. У 3553 работодателей в качестве сферы деятельности указана Разработка программного обеспечения;
5. У Яндекс имеется 485 вакансии в 16 городах-миллионниках

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [138]:
# текст запроса
query_6_1 = f'''SELECT count(*)
                FROM vacancies
                WHERE name ILIKE '%data%' OR name ILIKE '%данн%'
            '''

In [139]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
display(df_6_1)
print('Количество вакансий, которое имеет отношение к данным:', df_6_1['count'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2067247855.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


count
0   1771

Количество вакансий, которое имеет отношение к данным: 1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [140]:
# текст запроса
query_6_2 = f'''SELECT count(*)
                FROM vacancies
                -- Условия по ключевым словам для определения вакансий начального дата-сайентиста --
                WHERE (name ILIKE '%data scientist%'
                    OR name ILIKE '%data science%'
                    OR name ILIKE '%исследователь данных%'
                    OR name LIKE '%ML%'
                    OR name ILIKE '%machine learning%'
                    OR name ILIKE '%машинн%обучен%')
                    AND name NOT ILIKE '%html%'
                    AND (name ILIKE '%junior%'
                    OR experience ILIKE '%нет опыта%'
                    OR employment ILIKE '%стажировка%')
            '''

In [141]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
display(df_6_2)
print('Количество вакансий для начинающего дата-сайентиста:', df_6_2['count'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2437479008.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


count
0     51

Количество вакансий для начинающего дата-сайентиста: 51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?
** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [142]:
# текст запроса
query_6_3 = f'''SELECT count(*)
                FROM vacancies
                -- Условия по ключевым словам для определения вакансий дата-сайентиста с ключевыми навыками SQL или postgres  --
                WHERE (name ILIKE '%data scientist%'
                    OR name ILIKE '%data science%'
                    OR name ILIKE '%исследователь данных%'
                    OR name LIKE '%ML%'
                    OR name ILIKE '%machine learning%'
                    OR name ILIKE '%машинн%обучен%')
                    AND name NOT ILIKE '%html%'
                    AND (key_skills ILIKE '%SQL%'
                    OR key_skills ILIKE '%postgres%')
            '''

In [143]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
display(df_6_3)
print('Количество вакансий для дата-сайентиста, у которого в качестве ключевого навыка указан SQL или postgres:', df_6_3['count'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\1299210802.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


count
0    201

Количество вакансий для дата-сайентиста, у которого в качестве ключевого навыка указан SQL или postgres: 201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [144]:
# текст запроса
query_6_4 = f'''SELECT count(*)
                FROM vacancies
                -- Условия по ключевым словам для определения вакансий дата-сайентиста с ключевым навыком Python  --
                WHERE (name ILIKE '%data scientist%'
                    OR name ILIKE '%data science%'
                    OR name ILIKE '%исследователь данных%'
                    OR name LIKE '%ML%'
                    OR name ILIKE '%machine learning%'
                    OR name ILIKE '%машинн%обучен%')
                    AND name NOT ILIKE '%html%'
                    AND key_skills ILIKE '%python%'
            '''

In [145]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
display(df_6_4)
print('Количество вакансий дата-сайентиста, в которых в качестве ключевого навыка указан Python:', df_6_4['count'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2243834886.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


count
0    351

Количество вакансий дата-сайентиста, в которых в качестве ключевого навыка указан Python: 351


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [146]:
# текст запроса
query_6_5 = f'''SELECT
                    -- Определение количества ключевых навыков в вакансиях для DS --
                    round(avg(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1), 2) count_skills
                FROM vacancies
                WHERE 
                (name ILIKE '%data scientist%'
                OR name ILIKE '%data science%'
                OR name ILIKE '%исследователь данных%'
                OR (name NOT ILIKE '%HTML%' and name LIKE '%ML%')
                OR name ILIKE '%machine learning%'
                OR name ILIKE '%машинн%обучен%%')
                AND key_skills IS NOT NULL
            '''

In [147]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
display(df_6_5)
print('Количество ключевых навыков, которое в среднем указывают в вакансиях для DS:', df_6_5['count_skills'][0])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\3240752302.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


count_skills
0          6.41

Количество ключевых навыков, которое в среднем указывают в вакансиях для DS: 6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [148]:
# текст запроса
query_6_6 = f'''SELECT 
                    experience,
                    -- Определение средней зарплаты для DS -- 
                    ROUND(AVG(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)), 0) avg_salary
                FROM vacancies
                WHERE
                (name ILIKE '%data scientist%'
                OR name ILIKE '%data science%'
                OR name ILIKE '%исследователь данных%'
                OR (name  LIKE '%ML%' AND  name NOT iLIKE '%html%')
                OR name ILIKE '%machine learning%'
                OR name ILIKE '%машинн%обучен%')
                AND key_skills IS NOT NULL
                AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
                GROUP BY experience
                ORDER BY 1
            '''

In [149]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
display(df_6_6)
print('Зарплата в среднем, на которую может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет:', df_6_6['avg_salary'][2])

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2591348804.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    146704.0
2       От 3 до 6 лет    243115.0

Зарплата в среднем, на которую может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет: 243115.0


***

Из предметного анализа можно сделать выводы:
1. 1171 вакансии имеет отношение к данным;
2. 51 вакансия есть на платформе для начинающего дата-сайентиста;
3. Размещены на платформе 201 вакансия для дата-сайентиста, у которого в качестве ключевого навыка указан SQL или postgres;
4. Размещены на платформе 351 вакансия дата-сайентиста, в которых в качестве ключевого навыка указан Python;
5. Количество ключевых навыков, которое в среднем указывают в вакансиях для DS - 6.41;
6. Зарплата в среднем, на которую может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет - 243115 рублей.

Из всего вышеизложенного можно сделать общий вывод, что шанс устроится на работу дата-сайентистом выше у людей, которые имеют: опыт работы, в качестве ключевого навыка указан SQL или postgres и Python, а также количество ключевых навыков 6 и более.

# Дополнительные исследования данных

1. Из запроса ниже можно наблюдать, что самой востребованной вакансией является "системный администратор", затем "Програмист 1С" и "Бизнес аналитик"

In [150]:
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования
query_7_1 = f'''SELECT 
                    name,
                    count(*)
                FROM vacancies
                GROUP BY name
                ORDER BY 2 DESC, 1
            '''
df_7_1 = pd.read_sql_query(query_7_1, connection)
display(df_7_1)

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\2498332556.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_1 = pd.read_sql_query(query_7_1, connection)


name  count
0                                Системный администратор   1761
1                                         Программист 1С   1153
2                                        Бизнес-аналитик    488
3                       Специалист технической поддержки    457
4                                     Системный аналитик    432
...                                                  ...    ...
21218  Юрисконсульт/Специалист по защите персональных...      1
21219                                              Юрист      1
21220  Юрист- аналитик (127-ФЗ О несостоятельности / ...      1
21221                  Юрист-аналитик в Клиентский отдел      1
21222                 Юрист в сфере международного права      1

[21223 rows x 2 columns]

2. Из запроса ниже можно наблюдать, что самые популярные наборы требований к ключевым навыкам - ( Python, Machine Learning, Data Analysis, Numpy, Pandas )  и  ( Оценка рисков, MS PowerPoint, Статистический анализ, Управление рисками, Математическое моделирование, PD, LGD, EAD, SQL, Портфельные риски, кредитные риски )

In [151]:
query_7_2 = f'''SELECT
                    -- Преобразование ключевых навыков в одну строку с разделителем "," --
                    replace(key_skills, CHR(9), ', ') key_skills, 
                    count(*)
                FROM vacancies
                WHERE
                (name ILIKE '%data scientist%'
                OR name ILIKE '%data science%'
                OR name ILIKE '%исследователь данных%'
                OR (name  LIKE '%ML%' AND  name NOT iLIKE '%html%')
                OR name ILIKE '%machine learning%'
                OR name ILIKE '%машинн%обучен%')
                AND key_skills IS NOT NULL
                GROUP BY 1
                ORDER BY 2 DESC
            '''

df_7_2 = pd.read_sql_query(query_7_2, connection)
display(df_7_2)
print('Самые популярные требования к ключевым навыкам Data Science: ','(' ,df_7_2['key_skills'][0],')' , ' и ', '(', df_7_2['key_skills'][1], ')')

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\1371741016.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_2 = pd.read_sql_query(query_7_2, connection)


key_skills  count
0    Python, Machine Learning, Data Analysis, Numpy...     14
1    Оценка рисков, MS PowerPoint, Статистический а...     13
2    Qt, ООП, C++, MS Visual Studio, Разработка ком...      9
3    Python, machine learning, web security, sklear...      9
4    Git, ООП, Atlassian Jira, Java, Atlassian Conf...      7
..                                                 ...    ...
313  Python, Computer Vision, Tensorflow, PyTorch, ...      1
314  Python, C++, SQL, Автоматизация технологически...      1
315  Python, C++, Написание статей, Написание научн...      1
316  Python, C++, Написание статей, Написание научн...      1
317  Python, C++, Оптимизация кода, Анализ кода, Ма...      1

[318 rows x 2 columns]

Самые популярные требования к ключевым навыкам Data Science:  ( Python, Machine Learning, Data Analysis, Numpy, Pandas )  и  ( Оценка рисков, MS PowerPoint, Статистический анализ, Управление рисками, Математическое моделирование, PD, LGD, EAD, SQL, Портфельные риски, кредитные риски )


3. Из запроса ниже можно сделать вывод, что наиболее популярной вакансией в Москве являются "Бизнес-аналитик", "Аналитик", "Системный аналитик", что в общих чертах совпадает с тенденцие по всей стране за исключением порядка, например, среди всех вакансий на платформе лидирует "Системный администратор"

In [152]:
query_7_3 = f'''SELECT 
                    v.name,
                    count(v)
                FROM vacancies v
                JOIN areas a ON v.area_id = a.id
                WHERE a.name = 'Москва'
                GROUP BY v.name
                ORDER BY 2 DESC, 1
            '''

df_7_3 = pd.read_sql_query(query_7_3, connection)
display(df_7_3)

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\1714462997.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_3 = pd.read_sql_query(query_7_3, connection)


name  count
0                                Бизнес-аналитик    173
1                                       Аналитик     69
2                             Системный аналитик     59
3                        Системный администратор     50
4                                Аналитик данных     41
...                                          ...    ...
3889                Эксперт сопровождения Hadoop      1
3890     Эксперт сопровождения Hadoop (SberData)      1
3891  Юнити-разработчик Junior (Unity developer)      1
3892                                       Юрист      1
3893          Юрист со знанием английского языка      1

[3894 rows x 2 columns]

4. Из запроса ниже можно сделать вывод, что наибольшая средняя зарплата зависит от размера региона: чем он больше, тем выше средняя зарплата в рамках Российской Федерации

In [153]:
query_7_4 = f'''SELECT 
                    a.name,
                    -- Определение средней зарплаты в зависимости от региона -- 
                    ROUND(AVG(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0) avg_salary
                FROM vacancies v
                LEFT JOIN areas a ON v.area_id = a.id
                WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
                GROUP BY a.name
                ORDER BY 2 DESC
            '''
df_7_4 = pd.read_sql_query(query_7_4, connection)
display(df_7_4)

C:\Users\XmuriKot\AppData\Local\Temp\ipykernel_2192\1966489813.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_4 = pd.read_sql_query(query_7_4, connection)


name  avg_salary
0       Германия    452751.0
1       Болгария    324348.0
2        Испания    300000.0
3     Черногория    288589.0
4          Литва    263019.0
..           ...         ...
666        Урень     13800.0
667           Ош     12743.0
668       Кобрин     12389.0
669  Степногорск     12205.0
670     Шахтинск      9764.0

[671 rows x 2 columns]

# Общий вывод по проекту

В ходе анализа вакансий hh.ru было проделано следующее:
1. Было посчитано количество вакансий, работодателей, регионов, сфер деятельности;
2. Было посчитано количество вакансий в каждом регионе, количество вакансий с хотя бы одним заполнееым полем с зарплатой, средние значения нижней и верхней границы зарплатной вилки, количество вакансий для каждого сочетания типа графика и типа трудоустройства, а также количество вакансий по опыту работы;
3. Было определено какие работодатели находятся на первом и пятом месте по количеству вакансий, количество вакансий и работодателей для каждого региона, количество работодателей без указания сферы деятельности, количество работодателей с вакансией с указанием сферы деятельности "Разработка программного обеспечения", количество регионов, в которых каждый работодатель публикует свои вакансии
4. Было определено количество вакансий в сфере данных, количество вакансий для начинающих дата-сайентистов, количество вакансий для дата-сайентистов с ключевым навыком "SQL" или "postgres", количество вакансий с ключевым навыком "python", количество навыков в среднем для дата-сайентиста, зарплата в среднем для дата-сайентиста с опытом работы  от 3 до 6 лет.

Также было выполнено дополнительное исследование, где были определены: количество вакансий в зависимости от сферы деятельности
колчество вакансий в зависимости от набора ключевых навыков, количество вакансий в зависимости от названия в Москве, а также средняя зарплата в зависимости от региона.

Из всего анализа можно сделать выводы:
1. Что наибольшее количество вакансий размещено в Москве. Это связано с тем, что там проживает больше людей, чем в любом другом регионе;
2. Наибольшее количество вакансий у Яндекса, связано это с тем, что они охватили многие области, напримерц: поиск и портальные сервисы, электронная коммерция, сервисы объявлений, медиасервисы, сервисы каршеринга, онлайн-заказа такси и заказа еды, сервисы доставки, агрегаторы товаров и услуг, беспилотные автомобили, что влечет за собой необходимость в большом штате сотрудников.
3. Наиболее востребованны сотрудники с опытом работы от 1 до 3 лет. Связано это с тем, что найти людей с опытом работы более 3  лет намного сложнее, чем от 1 до 3 лет, а сотрудники без опыта работы нужны далеко не всем работодателям.
4. Наиболее востребованная вакансия - Системный администратор.

# Дополнительные исследования данных